feat: Function Dispatching (#53)
huff-language/huff-docs: src/tutorial/function-dispatching/README.md

# Function Dispatching

Function dispatching is something that is fundamental to any Huff contract. Unlike Solidity and Vyper; Huff does not abstract function dispatching. In this section we will go over how dispatching is performed in the other languages, and how you may go about it in Huff.

## What is the problem?

> In the evm, contracts are interacted with by sending messages to them. The ABI standard exists as a canonical way to encode these messages, handling how inputs to functions should be encoded. These strict rules are what allow contracts to understand each other. The ABI standard also tells the contract which function the message intends to interact with. This is done by encoding a 4 byte selector at the beginning of the message. This 4 byte selector is the `keccak` of a function's signature. If you have written interfaces in solidity you may not realize that you are just providing your current contract the ability to generate the 4 byte selectors of a foreign contract.

Reviewer: ```suggestion
In the evm, contracts are interacted with by sending messages to them. The ABI standard exists as a canonical way to encode these messages, handling how inputs to functions should be encoded. These strict rules are what allow contracts to understand each other. The ABI standard also tells the contract which function the message intends to interact with. This is done by encoding a 4 byte selector at the beginning of the message. This 4 byte selector is the first 4 bytes of the `keccak` of the function's signature string. For example, the keccak256 hash of the string "myFunction(address,uint256)" is `0x451c00ddd225deee9948ba5eca26042a5ea1cc0980e5a5fb0a057f90567af5e0`.  So the first 4 bytes `0x451c00dd` is what Solidity uses as the function signature. If you have written interfaces in solidity you may not realize that you are just providing your current contract the ability to generate the 4 byte selectors of a foreign contract.
```

The rest of this section will detail two types of dispatching, linear dispatching and binary search dispatching.

## Linear Dispatching

From reading the above, or from reading a Huff contract before you may have developed some intuition about how the simplest way to perform dispatching is - A linear lookup. This method will extract the function selector from the calldata message, then brute force compare it to every other function in the contract.

The example below shows what a linear dispatcher may look like for a standard ERC20 token:

```huff

> // Interface

Reviewer: these section headers (Interface, Metadata) are a bit opinionated and maybe unecessary for purposes of this example

Author: Good call; I'll remove them

For example:

```huff
#define function returnOne() public returns(uint256)
#define function returnTwo() public returns(uint256)

/// @notice returns the value one
#define macro RETURN_ONE() = {
    0x01 0x00 mstore
    0x20 0x00 return
}

/// @notice places the value 2 onto the stack
#define macro RETURN_TWO() = {
    0x02
}

#define macro MAIN() = {
    0x00 calldataload 0xE0 shr

    // Dispatcher
    dup1 __FUNC_SIG(returnOne) eq returnOneJump jumpi
    dup1 __FUNC_SIG(returnTwo) eq returnTwoJump jumpi

    // Macros
    returnTwo:
        RETURN_TWO()
    returnOne:
        RETURN_ONE()
}
```

In the above macro, regardless if what message is sent to it, it will ALWAYS return the value 1. As the macro `RETURN_TWO` does not terminate, it will roll over into the execution of `RETURN_ONE`, terminating with returning the value one. As there is no guard to protect against non valid function signatures any message, even if it matches nothing in the dispatcher will return the value 1.

> To protect against no valid function selector being found we can include `0x00 0x00 revert` in-between our dispatcher and our macro jump labels as so:

Reviewer: setting up for "inheritance" later 
```suggestion
One way we can handle no valid function selector being found, is by inserting `0x00 0x00 revert` in-between our dispatcher and our macro jump labels as so:
```

```huff
#define macro MAIN() = {
    0x00 calldataload 0xE0 shr

    dup1 __FUNC_SIG(returnOne) eq returnOneJump jumpi
    dup1 __FUNC_SIG(returnTwo) eq returnTwoJump jumpi

    0x00 0x00 revert

    returnTwo:
        RETURN_TWO()
    returnOne:
        RETURN_ONE()
}
```

> Please keep this behavior in mind when writing Huff contracts, especially when dealing with administration functions, running into a `SET_OWNER()` macro could allow anyone to take control of your contract.

Reviewer: We can just add that new section on inheritance later and leave this vague reference for now.
```suggestion
Please keep this behavior in mind when writing Huff contracts, especially when dealing with administration functions, running into a `SET_OWNER()` macro could allow anyone to take control of your contract.  There is a more advanced way of jumping over the dispatcher that is used with Huff "inheritance" pattern that we will discuss in another section.
```
